File: RecoLuminosity/LumiDB/python/idDealer.py
import coral
import nameDealer
class idDealer(object):
    """Manages the autoincremental ID values.\n
    Input: coral.schema object
    """
    def __init__( self , schema  ):
        self.__schema = schema
        self.__idTableColumnName = nameDealer.idTableColumnDefinition()[0]
        self.__idTableColumnType = nameDealer.idTableColumnDefinition()[1]
        
    def getIDColumnDefinition( self ):
        return (self.__idTableColumnName, self.__idTableColumnType)
    
    def getIDforTable( self, tableName ):
        """
        get the new id value to use for the given table
        """
        try:
            idtableName = nameDealer.idTableName(tableName)
            query = self.__schema.tableHandle(idtableName).newQuery()
            query.addToOutputList(self.__idTableColumnName)
            query.setForUpdate() #lock it
            cursor = query.execute()
            result = 0
            while ( cursor.next() ):
                result = cursor.currentRow()[self.__idTableColumnName].data()
            del query
            return result
        except Exception, e:
            raise Exception, str(e)

    def generateNextIDForTable( self, tableName ):
        """
        Set the nextID in the IDTableName to current id value + 1 .\n
        Input: ID table name.
        """
        try:
            idtableName = nameDealer.idTableName(tableName)
            tableHandle = self.__schema.tableHandle(idtableName)
            query = tableHandle.newQuery()
            query.addToOutputList(self.__idTableColumnName)
            query.setForUpdate() #lock it
            cursor = query.execute()
            result = 0
            while ( cursor.next() ):
                result = cursor.currentRow()[0].data()
            dataEditor = tableHandle.dataEditor()
            inputData = coral.AttributeList()
            dataEditor.updateRows('NEXTID = NEXTID+1','',inputData)
            del query            
            return result+1
        except Exception, e:
            raise Exception, str(e)

if __name__ == "__main__":
    fakeIDtableName='Fake_ID'
    svc=coral.ConnectionService()
    session=svc.connect('sqlite_file:fake.db')
    transaction=session.transaction()
    try:
        transaction.start(False)
        schema=session.nominalSchema()
        idor=idDealer(schema)
        if schema.existsTable(fakeIDtableName) is False:
          description=coral.TableDescription()
          description.setName(fakeIDtableName)
          description.setPrimaryKey(idor.getIDColumnDefinition()[0])        
          description.insertColumn(idor.getIDColumnDefinition()[0],idor.getIDColumnDefinition()[1])
          idtableHandle=schema.createTable(description)
          idtableHandle.privilegeManager().grantToPublic(coral.privilege_Select)
          inputData=coral.AttributeList()
          editor=idtableHandle.dataEditor()
          editor.rowBuffer(inputData)
          inputData[ idor.getIDColumnDefinition()[0] ].setData(0)
          editor.insertRow(inputData)
        idor.generateNextIDForTable('Fake')
        print idor.getIDforTable('Fake')
        transaction.commit()
        del session
    except coral.Exception,e:
        transaction.rollback()
        del session
    except Exception, e:
        print 'failed in unit test'
        print str(e)
        del session

File: RecoLuminosity/LumiProducer/src/idDealer.cc
#include "RecoLuminosity/LumiProducer/interface/idDealer.h"
#include "RelationalAccess/ISchema.h"
#include "RelationalAccess/ITable.h"
#include "RelationalAccess/IQuery.h"
#include "RelationalAccess/ICursor.h"
#include "RelationalAccess/ITableDataEditor.h"
#include "CoralBase/AttributeList.h"
#include "CoralBase/AttributeSpecification.h"
#include "CoralBase/AttributeList.h"
#include "CoralBase/Attribute.h"
#include "RecoLuminosity/LumiProducer/interface/LumiNames.h"
#include <sstream>
//#include <iostream>
lumi::idDealer::idDealer( coral::ISchema& schema):m_schema(schema),m_idtablecolumnName(lumi::LumiNames::idTableColumnName()),m_idtablecolumnType(lumi::LumiNames::idTableColumnType()){
}
unsigned long long lumi::idDealer::getIDforTable( const std::string& tableName ){
  std::string idtableName=lumi::LumiNames::idTableName(tableName);
  coral::IQuery* q=m_schema.tableHandle(idtableName).newQuery();
  q->addToOutputList(m_idtablecolumnName);
  q->setForUpdate(); //lock it
  coral::ICursor& cursor=q->execute();
  unsigned long long result=0;
  while ( cursor.next() ){
    const coral::AttributeList& row = cursor.currentRow();
    result = row[m_idtablecolumnName].data<unsigned long long>();
  }
  cursor.close();
  delete q;
  return result;
}
unsigned long long lumi::idDealer::generateNextIDForTable( const std::string& tableName,unsigned int interval ){
  std::string idtableName=lumi::LumiNames::idTableName(tableName);
  coral::IQuery* q=m_schema.tableHandle(idtableName).newQuery();
  q->addToOutputList(m_idtablecolumnName);
  unsigned long long r=0;
  coral::AttributeList myresult;
  myresult.extend(m_idtablecolumnName,typeid(unsigned long long));
  q->defineOutput(myresult);
  q->setForUpdate(); //lock it
  coral::ICursor& cursor=q->execute();
  while ( cursor.next() ){
    const coral::AttributeList& row = cursor.currentRow();
    r = row[m_idtablecolumnName].data<unsigned long long>();
  }
  coral::ITableDataEditor& dataEditor=m_schema.tableHandle(idtableName).dataEditor();
  coral::AttributeList inputData;
  //inputData.extend("interval",typeid(unsigned int));
  //inputData["interval"].data<unsigned int>()=interval;
  //dataEditor.updateRows(m_idtablecolumnName+"="+m_idtablecolumnName+"+1","",inputData);
  std::stringstream ss;
  ss<<interval;
  dataEditor.updateRows(m_idtablecolumnName+"="+m_idtablecolumnName+"+"+ss.str(),"",inputData);
  delete q;
  return r+interval;
}
